File: apps/api/app/schemas/milestone.py
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field


class MilestoneBase(BaseModel):
	sequence: int
	title: str
	description: str
	amount: Decimal


class MilestoneCreate(MilestoneBase):
	pass


class MilestoneUpdate(BaseModel):
	title: str | None = None
	description: str | None = None
	amount: Decimal | None = None


class MilestoneRead(MilestoneBase):
	model_config = ConfigDict(from_attributes=True, populate_by_name=True)

	id: str
	job_id: str = Field(alias="jobId")
	status: str
	submission_note: str | None = Field(default=None, alias="submissionNote")
	submitted_at: str | None = Field(default=None, alias="submittedAt")
	approved_at: str | None = Field(default=None, alias="approvedAt")
	disputed_at: str | None = Field(default=None, alias="disputedAt")
	resolved_at: str | None = Field(default=None, alias="resolvedAt")
	created_at: datetime = Field(alias="createdAt")

File: apps/web/types/milestone.ts
/**
 * Milestone type — aligned with context.md and the FastAPI Pydantic schema.
 *
 * IMPORTANT: amount is a string. Never use JS number for USDC values.
 */

export type MilestoneStatus =
  | "PENDING"
  | "FUNDED"
  | "IN_PROGRESS"
  | "PENDING_REVIEW"
  | "SUBMITTED"
  | "APPROVED"
  | "RELEASED"
  | "DISPUTED"
  | "RESOLVED";

export interface Milestone {
  id: string;
  jobId: string;
  /** 1-indexed ordering within the job */
  sequence: number;
  title: string;
  description: string;
  /** USDC amount as a string — never parse to float */
  amount: string;
  status: MilestoneStatus;
  /** Freelancer's delivery note set on submission */
  submissionNote: string | null;
  submittedAt: string | null;
  approvedAt: string | null;
  disputedAt: string | null;
  resolvedAt: string | null;
  createdAt: string;
}
